enhancements/single-node-developer-cluster-profile.md

```diff
@@ -0,0 +1,643 @@
+
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  namespace: openshift-console
+  name: downloads
+  annotations:
+    include.release.openshift.io/self-managed-high-availability: "true"
+    include.release.openshift.io/single-node-developer: "true"
+spec:
+  replicas: 2
+  selector:
+    # ...
+```
+
+In order to document that they have been purposely excluded from the
+profile, the following annotation, which is ignored by the CVO, will
+be added to all manifests not included in the profile:
+
+```
+include.release.openshift.io/single-node-developer: "false"
+```
+
+Once PRs adding these annotations have been merged to all the
+necessary component repositories, the snc.sh script will be updated to
+tell the OpenShift installer to use the 'single-node-developer'
+cluster profile, and to remove the code that deletes the components
+that are not included in the profile. At the conclusion of this
+initial phase, the somewhat simplified SNC image build tooling should
+produce a cluster equivalent to that currently produced, and all
+existing CRC tests should pass. In addition, the
+'single-node-developer' profile conformance CI job will be enabled at
+this point, as described below.
+
+From this point forward, any new component being added to OpenShift
+will not be included in CRC unless a manifest with the
+'single-node-developer' profile annotation specifying a value of
+"true" is included in the new component. The CRC team will need to
+engage with the OpenShift component team to help decide whether or not
+the new component should be included in the profile, and if so,
+whether it needs any special configuration for CRC. Once a decision
+has been reached, the annotation should be added to the component's
+manifests with a value indicating the result.
+
+### Phase Two
+
+In the second phase, various OpenShift components that are included in
+single node developer clusters will be updated to use specialized
+manifests in the 'single-node-developer' profile rather than the same
+manifest used for default clusters. These specialized manifests will
+configure the components for use in single node developer clusters,
+replacing the patching currently done post-deployment by the CRC build
+process, and potentially enabling additional reductions in resource
+requirements.
+
+Some of the manifest specializations expected to be included in this
+phase are shown in the following table:
+
+
+| Operator | Change |
+| :--- | :--- |
+| ingress | reduce the number of replicas of controller pod |
+| console | reduce the number of replicas of console pod |
+| console | reduce the number of replicas of downloads pod |
+| authentication | reduce the number of replicas of oauth pod |
+| lifecycle-manager | reduce the number of replicas of packageserver pod |
+| etcd | have a configuration with useUnsupportedUnsafeNonHANonProductionUnstableEtcd option |
+| monitoring | disable it but still create the CRDs (esp. PrometheusRule and ServiceMonitor) |
```

Comment:
I do not support disabling monitoring in the long run on CRC.  Monitoring of apps, and being able to test app monitoring, is a fundamental part of building applications on openshift.  The fact that it has been historically disabled in CRC is not an argument for current disablement.

Reply:
I (and I think the entire CRC team) completely agree on monitoring being needed by developers in the long run. But my understanding is that CRC cannot enable it without somehow significantly reducing its resource requirements. This enhancement helps enable future work where its configuration/implementation can be optimized for the CRC use case, potentially as part of phase 2 (and post 4.7). 

So what do you recommend for the shorter term? One option would be to include the monitoring-related manifests in the profile for phase 1, and continue to disable its resources via the SNC tooling as is done currently. I think that would allow CRC users who need monitoring and have sufficient memory to enable it on day 2. Any other options?

Reply:
We totally want monitoring enabled by default (now) but will work to get it going long term. As Bob stated, it was multi-GB RAM component and could not find a way to make it work with the requirements we have to try to fit on a developer's 16GB machine. If we can't find a way to get monitoring in initial go of this due to the continued size requirements, we can continue to disable it and provide instructions on how to use it. This has been working pretty well so far. Speaking with Ben Browning we have talked about work to reduce requirements for monitoring and we are working, though timing may be off. So we have a plan/approach to getting to it after the first phase.

Reply:
https://github.com/openshift/enhancements/pull/504#discussion_r527267112 covers how you can make it more efficient.

I would suggest working with single node produciton edge and pooling effort.  Monitoring team may also be able to suggest alternative resource efficiency wins.  Memory seems like the biggest impact to you, but single node production edge is about CPU, and there are approaches that reduce both (number of series scraped).

Reply:
In the upcoming update, monitoring is included in the profile in phase 1, but still disabled via the SNC tooling, and these sorts of optimizations are encouraged during phase 2. I agree pooling efforts with the production edge team on this makes sense.